crypto/ledger_secp256k1.go

```diff
@@ -49,47 +48,47 @@ type (
 //
 // CONTRACT: The ledger device, ledgerDevice, must be loaded and set prior to
 // any creation of a PrivKeyLedgerSecp256k1.
-func NewPrivKeyLedgerSecp256k1(path DerivationPath) (tmcrypto.PrivKey, error) {
-	if discoverLedger == nil {
-		return nil, errors.New("no Ledger discovery function defined")
-	}
-
-	device, err := discoverLedger()
+func NewPrivKeyLedgerSecp256k1(path hd.BIP44Params) (tmcrypto.PrivKey, error) {
+	device, err := getLedgerDevice()
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to create PrivKeyLedgerSecp256k1")
+		return nil, err
 	}
+	defer warnIfErrors(device.Close)
 
-	pkl := &PrivKeyLedgerSecp256k1{Path: path, ledger: device}
-
-	pubKey, err := pkl.getPubKey()
+	pubKey, err := getPubKey(device, path)
 	if err != nil {
 		return nil, err
 	}
 
-	pkl.CachedPubKey = pubKey
-	return pkl, err
+	return PrivKeyLedgerSecp256k1{pubKey, path}, nil
 }
 
 // PubKey returns the cached public key.
 func (pkl PrivKeyLedgerSecp256k1) PubKey() tmcrypto.PubKey {
 	return pkl.CachedPubKey
 }
 
+// Sign returns a secp256k1 signature for the corresponding message
+func (pkl PrivKeyLedgerSecp256k1) Sign(message []byte) ([]byte, error) {
```

Comment:
This doesn't actually operate on the receiver? Can we turn this into a vanilla function?

Reply:
This is a common interface defined in tendermint/crypto for all private keys.

https://github.com/tendermint/tendermint/blob/4d7b29cd8f8ef998cf8f7141f3a100b0a4fa718d/crypto/crypto.go#L31

It applies also to Secp256k1 and ed25519:
https://github.com/tendermint/tendermint/blob/4d7b29cd8f8ef998cf8f7141f3a100b0a4fa718d/crypto/ed25519/ed25519.go#L55
https://github.com/tendermint/tendermint/blob/master/crypto/secp256k1/secp256k1.go#L48

The refactoring resulted in a slightly different implementation but I had to keep the same interface as most of the other code relies on being able to call Sign on PrivKey interface.